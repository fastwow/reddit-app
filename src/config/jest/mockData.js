export const postsResponse = {
  children: [
    {
      data: {
        url: 'https://v.redd.it/1zhxx58rhp421',
        id: '98vj9e',
        thumbnail: 'https://b.thumbs.redditmedia.com/NFJ5zzCUnxTgAl71EkjSSCePc7wtcCLqTdbYgrwJkdU.jpg',
        title: 'Blinks of affection',
      },
    },
    {
      data: {
        url: 'https://i.imgur.com/GsoJi0d.gifv',
        id: 'a6q94b',
        thumbnail: 'https://b.thumbs.redditmedia.com/cjYmtjKiy9OT1y43RFMuTkcH_WmTM66RsSuN7c6xkhA.jpg',
        title: 'This church looks like a gingerbread house.',
      },
    },
  ],
};

export const posts = postsResponse.children.map(c => c.data);
