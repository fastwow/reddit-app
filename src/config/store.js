/* global __DEV__ */
import {createStore, applyMiddleware, compose} from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import rootReducer from '../reducers';
import storage from 'redux-persist/lib/storage';
import {persistReducer} from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';

let middleware = [thunk];

if (__DEV__) {
  const reduxImmutableStateInvariant = require('redux-immutable-state-invariant').default();
  middleware = [...middleware, reduxImmutableStateInvariant, logger];
} else {
  middleware = [...middleware];
}

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['favorites'],
  debug: __DEV__,
  stateReconciler: autoMergeLevel2,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export default function configureStore(initialState) {
  const enhancer = compose(
    applyMiddleware(...middleware),
  );
  return createStore(persistedReducer, initialState, enhancer);
}
