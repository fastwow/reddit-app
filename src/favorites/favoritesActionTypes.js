export const ADD_TO_FAVORITE = 'add_to_favorite';
export const REMOVE_FROM_FAVORITE = 'remove_from_favorite';
