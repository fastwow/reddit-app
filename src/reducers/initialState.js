export default {
  posts: {isLoading: false, isRefreshing: false, errorMessage: '', posts: [], after: ''},
  favorites: {posts: []},
  filter: {searchTerm: ''},
};
