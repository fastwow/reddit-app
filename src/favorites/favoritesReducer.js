import initialState from './favoritesActionsInitialState';
import {ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE} from './favoritesActionTypes';

export default function (state = initialState, action) {
  switch (action.type) {
    case ADD_TO_FAVORITE:
      return {
        ...state,
        posts: state.posts.concat(action.data),
      };
    case REMOVE_FROM_FAVORITE:
      return {
        ...state,
        posts: state.posts.filter(item => action.data.id !== item.id),
      };
    default:
      return state;
  }
}
