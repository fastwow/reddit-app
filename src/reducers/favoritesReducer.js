import initialState from './initialState';
import {ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE} from '../actions/types';

export default function (state = initialState.favorites, action) {
  switch (action.type) {
    case ADD_TO_FAVORITE:
      return {
        ...state,
        posts: state.posts.concat(action.data),
      };
    case REMOVE_FROM_FAVORITE:
      return {
        ...state,
        posts: state.posts.filter(item => action.data.id !== item.id),
      };
    default:
      return state;
  }
}
