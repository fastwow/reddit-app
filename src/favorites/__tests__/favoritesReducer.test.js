import initialState from '../favoritesActionsInitialState';
import favoritesReducer from '../favoritesReducer';
import {ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE} from '../favoritesActionTypes';
import {posts} from '../../shared/config/mockData';

describe('favoritesReducer', () => {
  it(`returns the same state for favorites reducer on an unhandled action`, () => {
    expect(favoritesReducer(initialState, {type: '_NULL'})).toMatchSnapshot();
  });

  it(`handles ${ADD_TO_FAVORITE} action`, () => {
    expect(favoritesReducer({...initialState, posts: [posts[0]]}, {
      data: posts[1],
      type: ADD_TO_FAVORITE,
    })).toMatchSnapshot();
  });

  it(`handles ${REMOVE_FROM_FAVORITE} action`, () => {
    expect(favoritesReducer({...initialState, posts}, {
      data: posts[1],
      type: REMOVE_FROM_FAVORITE,
    })).toMatchSnapshot();
  });
});
