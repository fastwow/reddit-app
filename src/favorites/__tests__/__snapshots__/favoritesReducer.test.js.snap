// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`favoritesReducer handles add_to_favorite action 1`] = `
Object {
  "favorites": Object {
    "posts": Array [],
  },
  "posts": Array [
    Object {
      "id": "98vj9e",
      "thumbnail": "https://b.thumbs.redditmedia.com/NFJ5zzCUnxTgAl71EkjSSCePc7wtcCLqTdbYgrwJkdU.jpg",
      "title": "Blinks of affection",
      "url": "https://v.redd.it/1zhxx58rhp421",
    },
    Object {
      "id": "a6q94b",
      "thumbnail": "https://b.thumbs.redditmedia.com/cjYmtjKiy9OT1y43RFMuTkcH_WmTM66RsSuN7c6xkhA.jpg",
      "title": "This church looks like a gingerbread house.",
      "url": "https://i.imgur.com/GsoJi0d.gifv",
    },
  ],
}
`;

exports[`favoritesReducer handles remove_from_favorite action 1`] = `
Object {
  "favorites": Object {
    "posts": Array [],
  },
  "posts": Array [
    Object {
      "id": "98vj9e",
      "thumbnail": "https://b.thumbs.redditmedia.com/NFJ5zzCUnxTgAl71EkjSSCePc7wtcCLqTdbYgrwJkdU.jpg",
      "title": "Blinks of affection",
      "url": "https://v.redd.it/1zhxx58rhp421",
    },
  ],
}
`;

exports[`favoritesReducer returns the same state for favorites reducer on an unhandled action 1`] = `
Object {
  "favorites": Object {
    "posts": Array [],
  },
}
`;
